apps/settings/js/modules/bluetooth/bluetooth_context.js

```diff
@@ -0,0 +1,852 @@
+     * @memberOf BluetoothContext
+     * @param {Object BluetoothAdapter} adapter
+     */
+    _refreshPairedDevicesInfo: function btc__refreshPairedDevicesInfo(adapter) {
+      var pairedDevices = adapter.getPairedDevices();
+      this.debug('pairedDevices.length = ' + pairedDevices.length);
+
+      // reset paired devices list
+      this._pairedDevices.reset([]);
+
+      // refresh properties about paired devices
+      if (pairedDevices.length === 0) {
+        // reset the name
+        this.firstPairedDeviceName = '';
+        // update property 'hasPairedDevice'
+        this.hasPairedDevice = false;
+      } else {
+        // save paired devices in list
+        this._getBluetoothDevice().then((BtDevice) => {
+          for (var i in pairedDevices) {
+            // create observable BtDevice
+            var observableBtDevice = BtDevice(pairedDevices[i]);
+            // push device in devices list with observable object
+            this._pairedDevices.push(observableBtDevice);
+          }
+        });
+
+        // copy for sorting
+        var paired = pairedDevices.slice();
+        paired.sort(function(a, b) {
+          return a.name > b.name;
+        });
+        // set the name
+        this.firstPairedDeviceName = paired[0].name;
+        // update property 'hasPairedDevice'
+        this.hasPairedDevice = true;
+      }
+
+      this.numberOfPairedDevices = pairedDevices.length;
+    },
+
+    /**
+     * Since we have three properties which are relative with hardware status,
+     * update them via this method here.
+     * Update status, enabled, settings key 'bluetooth.enabled'
+     * The input state would be {'disabled', 'disabling', 'enabled', 'enabling'}
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {String} state
+     */
+    _updateStatus: function btc__updateStatus(state) {
+      // Wrapper state to enabled/disabled toggle state.
+      var enabled;
+      if (state === 'enabled' || state === 'disabling') {
+        enabled = true;
+      } else if (state === 'disabled' || state === 'enabling') {
+        enabled = false;
+      }
+
+      // Sync with settings key
+      SettingsCache.getSettings((results) => {
+        var btEnabled = results['bluetooth.enabled'];
+        if (btEnabled !== enabled) {
+          settings.createLock().set({'bluetooth.enabled': enabled});
+        }
+
+        // Update state
+        this.state = state;
+        this.debug('_updateStatus(): set state = ' + state);
+
+        // Update enabled
+        this.enabled = enabled;
+        this.debug('_updateStatus(): set enabled = ' + enabled);
+      });
+    },
+
+    /**
+     * Set Bluetooth enable/disable.
+     *
+     * @access public
+     * @memberOf BluetoothContext
+     * @param {Boolean} enabled
+     */
+    setEnabled: function btc_setEnabled(enabled) {
+      this.debug('setEnabled(): request set enabled = ' + enabled);
+      if ((this.enabled === enabled) ||
```

Comment:
Please return a promise for public async functions. Take this function for example, we may return `Promise.reject(reason)` for this._defaultAdapter is not available or the state is still in transition. And return `Promise.resolve()` if there is nothing to do.